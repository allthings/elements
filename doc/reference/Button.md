<!-- 
This is an auto-generated markdown. 
You can change it in "src/Button/Button.tsx" and run build:docs to update this file.
-->
# Button
Buttons make common actions immediately visible and easy to perform with one
click or tap. They can be used for any type of action, including navigation.

You can use two different looks for the button: primary and
secondary. Primary is the default type, so there's no need to explicitly
define it.

```example
<ThemeProvider>
   <Button>Hello you</Button>
</ThemeProvider>
```

To have an icon as button-label, just add the icon-component as children.

```example
<Button type="submit">
    <View direction="row">
      Hello with icon
      <View style={{ marginLeft: 10 }}>
        <Icon name="send-filled" size="xs" color="white" />
      </View>
    </View>
  </Button>
```
## Usage
| Name        | Type           | Description  |
| ----------- |:--------------:| ------------:|
|secondary|boolean|If the button is used for a secondary option
|onClick|((event: MouseEvent<Element, MouseEvent>) => void)|
|type|"reset", "button", "submit"|Type of the button (deprecated)
|color|string|
|disabled|boolean|
|backgroundColor|string|Color of the button, theme primary color by default
|name|string|Name of the
