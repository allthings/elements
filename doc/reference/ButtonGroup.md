<!-- 
This is an auto-generated markdown. 
You can change it in "src/ButtonGroup/ButtonGroup.tsx" and run build:docs to update this file.
-->
# ButtonGroup
ButtonGroup groups multiple buttons and keeps the spacing consistent.

```example
<ButtonGroup>
   <Button secondary>Cancel</Button>
   <Button>Accept</Button>
</ButtonGroup>
```
## Usage
| Name        | Type           | Description  |
| ----------- |:--------------:| ------------:|
|htmlElement|string|
|alignH|"none", "start", "center", "end", "space-around", "space-between"|horizontal alignment
|alignV|"none", "start", "center", "end", "stretch"|vertical alignment
|direction|"row", "column", "row-reverse", "column-reverse"|direction
|fill|boolean|Passing true, will make the view fill out available space
|wrap|"wrap", "inherit", "initial", "nowrap", "wrap-reverse"|Defining how children will wrap
|flex|number, "flex", "none", "initial", "nogrow", "grow", "auto", "noshrink"|Flex values, can be 5, 10, 15 ... 100 or 33, 66
