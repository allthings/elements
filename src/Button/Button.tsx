import { css } from 'glamor'
import { color as col, lightness } from 'kewler'
import React, { HTMLProps, MouseEvent } from 'react'
import { createTextStyles } from '../Text'
import Theme from '../Theme'
import { color } from '../utils/propTypes/color'

const baseStyle = {
  position: 'relative',
  padding: '10px 16px',
  borderRadius: '2px',
  border: 'none',
  userSelect: 'none',
  outline: 'none',
}

const primary = ({
  color,
  disabled,
  disabledBackgroundColor,
  backgroundColor,
}: {
  color: color
  disabled: boolean
  disabledBackgroundColor: color
  backgroundColor: color
}) => ({
  color,
  background: disabled ? disabledBackgroundColor : backgroundColor,
})

const secondary = ({
  disabled,
  disabledColor,
  backgroundColor,
}: {
  disabled: boolean
  disabledColor: color
  backgroundColor: color
}) => ({
  background: 'transparent',
  color: disabled ? disabledColor : backgroundColor,
  margin: 2,
})

function styles(
  backgroundColor: string,
  color: string,
  isDisabled: boolean,
  isSecondary: boolean,
) {
  const props = {
    color,
    disabled: isDisabled,
    disabledBackgroundColor: 'lightGrey',
    disabledColor: 'darkgray',
    backgroundColor,
  }
  return css({
    ...baseStyle,
    ...(isSecondary ? secondary(props) : primary(props)),
    cursor: isDisabled ? 'not-allowed' : 'pointer',
    transition: '250ms ease-in-out',
    ':focus': {
      outline: 'none',
    },
    ':hover': isSecondary
      ? !isDisabled && { color: col(backgroundColor, lightness(-20)) }
      : {
          color,
          background: isDisabled
            ? props.disabledBackgroundColor
            : backgroundColor.indexOf('#') !== -1
            ? col(backgroundColor, lightness(-10))
            : backgroundColor,
        },
  })
}

interface IButtonProps extends HTMLProps<HTMLButtonElement> {
  /** If the button is used for a secondary option */
  readonly secondary?: boolean
  readonly onClick?: (event: MouseEvent) => void
  /** Type of the button (deprecated) */
  readonly type?: 'reset' | 'button' | 'submit'
  readonly color?: color
  readonly disabled?: boolean
  /** Color of the button, theme primary color by default */
  readonly backgroundColor?: string
  /** Name of the */
  readonly name?: string
}

/**
 * Buttons make common actions immediately visible and easy to perform with one
 * click or tap. They can be used for any type of action, including navigation.
 *
 * You can use two different looks for the button: primary and
 * secondary. Primary is the default type, so there's no need to explicitly
 * define it.
 *
 * ```example
 * <ThemeProvider>
 *   <Button>Hello you</Button>
 * </ThemeProvider>
 * ```
 *
 * To have an icon as button-label, just add the icon-component as children.
 *
 * ```example
 * <Button type="submit">
 *    <View direction="row">
 *      Hello with icon
 *      <View style={{ marginLeft: 10 }}>
 *        <Icon name="send-filled" size="xs" color="white" />
 *      </View>
 *    </View>
 *  </Button>
 * ```
 */
const Button = ({
  children,
  type = 'button',
  disabled = false,
  backgroundColor = 'primary',
  color = 'white',
  secondary = false,
  onClick,
  name,
  ...restProps
}: IButtonProps) => (
  <Theme>
    {({ colorize }) => (
      <button
        type={type}
        {...css(
          styles(
            colorize(backgroundColor),
            colorize(color),
            disabled,
            secondary,
          ),
        )}
        {...restProps}
        {...createTextStyles({ size: 'l' })}
        name={name || type}
        onClick={disabled ? undefined : onClick}
      >
        {children}
      </button>
    )}
  </Theme>
)

export default Button
