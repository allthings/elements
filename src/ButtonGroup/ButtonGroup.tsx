import * as React from 'react'
import View, { IViewProps } from '../View'
import { css } from 'glamor'

const style = css({
  '> *': {
    marginRight: 10,
  },
  '> *:last-child': {
    marginRight: 0,
  },
})

/**
 * ButtonGroup groups multiple buttons and keeps the spacing consistent.
 *
 * ```example
 * <ButtonGroup>
 *   <Button secondary>Cancel</Button>
 *   <Button>Accept</Button>
 * </ButtonGroup>
 * ```
 */
const ButtonGroup: React.FC<IViewProps> = ({ children, ...props }) => (
  <View direction="row" {...style} {...props}>
    {React.Children.map(children, child => (
      <View>{child}</View>
    ))}
  </View>
)

export default ButtonGroup
