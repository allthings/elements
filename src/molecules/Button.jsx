import PropTypes from 'prop-types'
import React from 'react'
import { css } from 'glamor'
import Text from '../atoms/Text'
import { withTheme } from '../behaviour/ThemeProvider'
import { color, colorCode } from '../propTypes/color'
import { color as col, lightness } from 'kewler'

const baseStyle = {
  position: 'relative',
  padding: '8px 14px',
  borderRadius: '2px',
  userSelect: 'none',
  outline: 'none',
  border: 'none',
}

function styles(
  backgroundColor,
  color,
  disabled,
  disabledColor,
  disabledBackgroundColor
) {
  return css({
    ...baseStyle,
    background: disabled ? disabledBackgroundColor : backgroundColor,
    color: disabled ? disabledColor : color,
    cursor: disabled ? 'not-allowed' : 'pointer',
    transition: '250ms ease-in-out',
    ':focus': {
      outline: 'none',
    },
    ':hover': {
      background: disabled
        ? disabledBackgroundColor
        : backgroundColor.includes('rgb')
        ? backgroundColor
        : col(backgroundColor, lightness(-10)),
    },
  })
}

/**
 * Buttons make common actions immediately visible and easy to perform with one
 * click or tap. They can be used for any type of action, including navigation.
 *
 * You can use two different looks for the button: primary and
 * secondary. Primary is the default type, so there's no need to explicitly
 * define it.
 *
 * ```example
 * <ThemeProvider>
 *   <Button>Hello you</Button>
 * </ThemeProvider>
 * ```
 *
 * To have an icon as button-label, just add the icon-component as children.
 *
 * ```example
 * <Button type="submit">
 *    <View direction="row">
 *      Hello with icon
 *      <View style={{ marginLeft: 10 }}>
 *        <Icon name="send-filled" size="xs" color="white" />
 *      </View>
 *    </View>
 *  </Button>
 * ```
 */
class Button extends React.Component {
  static propTypes = {
    /** Just text most of the time */
    children: PropTypes.node.isRequired,
    /** Called when the button is clicked */
    onClick: PropTypes.func,
    /** Type of the button (deprecated) */
    type: PropTypes.oneOf(['reset', 'button', 'submit']),
    /** Disable button state to indicate it's not touchable */
    disabled: PropTypes.bool,
    /** Color of the button, theme primary color by default */
    backgroundColor: PropTypes.string,
    /** Textcolor of the button (deprecated) */
    color: color,
    /** Textcolor when button is disabled (deprecated) */
    disabledColor: PropTypes.string,
    /** Color when button is disabled (deprecated) */
    disabledBackgroundColor: PropTypes.string,
    /** Pass your own css (deprecated) */
    css: PropTypes.object,
  }

  handleClick = e => {
    if (!this.props.disabled) {
      this.props.onClick && this.props.onClick(e)
    } else {
      e.preventDefault()
    }
  }

  render() {
    const {
      children,
      type,
      disabled,
      backgroundColor,
      color,
      disabledColor,
      disabledBackgroundColor,
      css: cssProp,
      ...restProps
    } = this.props

    const allStyles = css(
      styles(
        colorCode(backgroundColor),
        color,
        disabled,
        disabledColor,
        disabledBackgroundColor
      ),
      cssProp
    )

    return (
      <button
        type={type}
        {...allStyles}
        {...restProps}
        name={restProps.name || type || null}
        onClick={this.handleClick}
      >
        {typeof children === 'string' ? (
          <Text color={color}>{children}</Text>
        ) : (
          children
        )}
      </button>
    )
  }
}

Button.defaultProps = {
  type: 'button',
  disabled: false,
  color: 'white',
  backgroundColor: 'purple',
  disabledColor: 'darkgray',
  disabledBackgroundColor: 'lightGray',
}

const mapThemeToProps = (theme, props) => ({
  backgroundColor: props.backgroundColor || theme.primary,
})

export default withTheme(mapThemeToProps)(Button)
