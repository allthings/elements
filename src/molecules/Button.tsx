import React, { HTMLProps, MouseEvent, FunctionComponent } from 'react'
import { css } from 'glamor'
import { createTextStyles } from '../atoms/Text'
import { color } from '../propTypes/color'
import { color as col, lightness } from 'kewler'
import Theme from '../behaviour/Theme'

const baseStyle = {
  position: 'relative',
  padding: '10px 16px',
  borderRadius: '2px',
  border: 'none',
  userSelect: 'none',
  outline: 'none',
}

const primary = ({
  color,
  disabled,
  disabledBackgroundColor,
  backgroundColor,
}: {
  color: color
  disabled: boolean
  disabledBackgroundColor: color
  backgroundColor: color
}) => ({
  color,
  background: disabled ? disabledBackgroundColor : backgroundColor,
})

const secondary = ({
  disabled,
  disabledColor,
  backgroundColor,
}: {
  disabled: boolean
  disabledColor: color
  backgroundColor: color
}) => ({
  background: 'transparent',
  color: disabled ? disabledColor : backgroundColor,
  margin: 2,
})

function styles(
  backgroundColor: string,
  color: string,
  isDisabled: boolean,
  isSecondary: boolean,
) {
  const props = {
    color,
    disabled: isDisabled,
    disabledBackgroundColor: 'lightGrey',
    disabledColor: 'darkgray',
    backgroundColor,
  }
  return css({
    ...baseStyle,
    ...(isSecondary ? secondary(props) : primary(props)),
    cursor: isDisabled ? 'not-allowed' : 'pointer',
    transition: '250ms ease-in-out',
    ':focus': {
      outline: 'none',
    },
    ':hover': isSecondary
      ? !isDisabled && { color: col(backgroundColor, lightness(-20)) }
      : {
          color,
          background: isDisabled
            ? props.disabledBackgroundColor
            : backgroundColor.indexOf('#') !== -1
            ? col(backgroundColor, lightness(-10))
            : backgroundColor,
        },
  })
}

interface IButtonProps extends HTMLProps<HTMLButtonElement> {
  /** If the button is used for a secondary option */
  secondary?: boolean
  onClick?: (event: MouseEvent) => void
  /** Type of the button (deprecated) */
  type?: 'reset' | 'button' | 'submit'
  color?: color
  disabled?: boolean
  /** Color of the button, theme primary color by default */
  backgroundColor?: string
  /** Name of the */
  name?: string
}

/**
 * Buttons make common actions immediately visible and easy to perform with one
 * click or tap. They can be used for any type of action, including navigation.
 *
 * You can use two different looks for the button: primary and
 * secondary. Primary is the default type, so there's no need to explicitly
 * define it.
 *
 * ```example
 * <ThemeProvider>
 *   <Button>Hello you</Button>
 * </ThemeProvider>
 * ```
 *
 * To have an icon as button-label, just add the icon-component as children.
 *
 * ```example
 * <Button type="submit">
 *    <View direction="row">
 *      Hello with icon
 *      <View style={{ marginLeft: 10 }}>
 *        <Icon name="send-filled" size="xs" color="white" />
 *      </View>
 *    </View>
 *  </Button>
 * ```
 */
const Button: FunctionComponent<IButtonProps> = ({
  children,
  type = 'button',
  disabled = false,
  backgroundColor = 'primary',
  color = 'white',
  secondary = false,
  onClick,
  name,
  ...restProps
}) => (
  <Theme>
    {({ colorize }) => (
      <button
        type={type}
        {...css(
          styles(
            colorize(backgroundColor),
            colorize(color),
            disabled,
            secondary,
          ),
        )}
        {...restProps}
        {...createTextStyles({ size: 'l' })}
        name={name || type}
        onClick={disabled ? undefined : onClick}
      >
        {children}
      </button>
    )}
  </Theme>
)

export default Button
