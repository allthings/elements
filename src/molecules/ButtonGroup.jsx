import React from 'react'
import View from '../atoms/View'
import { css } from 'glamor'
import PropTypes from 'prop-types'

const style = css({
  '> *': {
    marginRight: 10,
  },
  '> *:last-child': {
    marginRight: 0,
  },
})

/**
 * ButtonGroup groups multiple buttons and keeps the spacing consistent.
 *
 * ```example
 * <ButtonGroup>
 *   <Button secondary>Cancel</Button>
 *   <Button>Accept</Button>
 * </ButtonGroup>
 * ```
 */
class ButtonGroup extends React.Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
  }

  render() {
    const { children } = this.props

    return (
      <View direction="horizontal" {...style}>
        {React.Children.map(children, child => (
          <View>{child}</View>
        ))}
      </View>
    )
  }
}

export default ButtonGroup
