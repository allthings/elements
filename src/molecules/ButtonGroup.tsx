import React, { FunctionComponent } from 'react'
import View, { IViewProps } from '../atoms/View'
import { css } from 'glamor'
import Button from './Button'

const style = css({
  '> *': {
    marginRight: 10,
  },
  '> *:last-child': {
    marginRight: 0,
  },
})

/**
 * ButtonGroup groups multiple buttons and keeps the spacing consistent.
 *
 * ```example
 * <ButtonGroup>
 *   <Button secondary>Cancel</Button>
 *   <Button>Accept</Button>
 * </ButtonGroup>
 * ```
 */
const ButtonGroup: FunctionComponent<IViewProps> = ({
  children,
  ...props
}: {
  children: [typeof Button][]
}) => (
  <View direction="row" {...style} {...props}>
    {React.Children.map(children, child => (
      <View>{child}</View>
    ))}
  </View>
)

export default ButtonGroup
