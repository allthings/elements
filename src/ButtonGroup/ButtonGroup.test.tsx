import * as React from 'react'
import ButtonGroup from './ButtonGroup'
import Button from '../Button'
import renderer from 'react-test-renderer'

describe('ButtonGroup', () => {
  it('should render button group', () => {
    const tree = renderer
      .create(
        <ButtonGroup>
          <Button secondary>Cancel</Button>
          <Button>Accept</Button>
        </ButtonGroup>,
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })
})
