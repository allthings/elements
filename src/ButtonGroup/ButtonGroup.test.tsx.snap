// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonGroup should render button group 1`] = `
.css-0> *,
[data-css-0]> * {
  margin-right: 10px;
}

.css-0> *:last-child,
[data-css-0]> *:last-child {
  margin-right: 0;
}

.css-1,
[data-css-1] {
  box-sizing: border-box;
  align-content: stretch;
  align-items: stretch;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  flex: 0 0 auto;
  -webkit-align-content: stretch;
  -webkit-box-align: stretch;
  -webkit-align-items: stretch;
  -webkit-box-orient: horizontal;
  -webkit-box-direction: normal;
  -webkit-flex-direction: row;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -webkit-flex: 0 0 auto;
}

.css-2,
[data-css-2] {
  box-sizing: border-box;
  flex: 0 0 auto;
  -webkit-flex: 0 0 auto;
}

.css-3,
[data-css-3] {
  position: relative;
  padding: 10px 16px;
  border-radius: 2px;
  border: none;
  user-select: none;
  outline: none;
  background: transparent;
  color: #3598db;
  margin: 2px;
  cursor: pointer;
  transition: 250ms ease-in-out;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  -webkit-transition: 250ms ease-in-out;
  -moz-transition: 250ms ease-in-out;
}

.css-3:focus,
[data-css-3]:focus {
  outline: none;
}

.css-3:hover,
[data-css-3]:hover {
  color: #1a6193;
}

.css-6,
[data-css-6] {
  display: inline;
  font-family: "Open Sans", Helvetica, Arial, sans-serif;
  font-weight: 400;
  font-size: 14px;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

.css-7,
[data-css-7] {
  position: relative;
  padding: 10px 16px;
  border-radius: 2px;
  border: none;
  user-select: none;
  outline: none;
  color: white;
  background: #3598db;
  cursor: pointer;
  transition: 250ms ease-in-out;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  -webkit-transition: 250ms ease-in-out;
  -moz-transition: 250ms ease-in-out;
}

.css-7:focus,
[data-css-7]:focus {
  outline: none;
}

.css-7:hover,
[data-css-7]:hover {
  color: white;
  background: #227dbf;
}

<div
  data-css-1=""
  data-css-0=""
>
  <div
    data-css-2=""
  >
    <button
      data-css-6=""
      data-css-3=""
      name="button"
      type="button"
    >
      Cancel
    </button>
  </div>
  <div
    data-css-2=""
  >
    <button
      data-css-6=""
      data-css-7=""
      name="button"
      type="button"
    >
      Accept
    </button>
  </div>
</div>
`;
